### PR TITLE
Ananyamukh6/psnr

### PR DESCRIPTION
**Pull request by ananyamukh6**
Implement PSNR loss

### CODE REVIEW

File: kornia/losses/psnr.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    def __init__(self, max_val) -> None:
+        super(PSNR, self).__init__()
+        self.max_val = max_val
+
+    def forward(  # type: ignore
+            self, signal: torch.Tensor, approximation: torch.Tensor) -> torch.Tensor:
+        return psnr(signal, approximation, self.max_val)
+
+
+def psnr(signal: torch.Tensor, approximation: torch.Tensor, max_val) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/losses/psnr.py]: `max_val` should typed as tensor

Comment (edgarriba) [inline kornia/losses/psnr.py]: In order to be consistent with PyTorch I would rename the function / module name as `psnr_loss` / `PSNRLoss` . In addition, for same reason as before rename `signal -> input` , `approximation -> target`, we can describe in the documentation what's the meaning of each variable.

File: kornia/losses/psnr.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    r"""Function that computes PSNR
+
+    See :class:`~kornia.losses.PSNR` for details.
+    """
+    if not torch.is_tensor(signal) or not torch.is_tensor(approximation):
+        raise TypeError(f"Expected 2 torch tensors but got {type(signal)} and {type(approximation)}")
+    if signal.shape != approximation.shape:
+        raise TypeError(f"Expected tensors of equal shapes, but got {signal.shape} and {approximation.shape}")
+    mse_val = mse_loss(signal, approximation, reduction='mean')
```
Comment (edgarriba) [inline kornia/losses/psnr.py]: similar to https://github.com/arraiyopensource/kornia/blob/master/kornia/losses/ssim.py#L135
we should support different reduction types: `'none' | 'mean' | 'sum'`

File: kornia/losses/psnr.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    Examples:
+        >>> kornia.losses.psnr(torch.ones(1), 1.2*torch.ones(1), 2)
+        tensor(20.0000) # 10 * log(4/((1.2-1)**2)) / log(10)
+
+    reference:
+        https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio#Definition
+    """
+
+    def __init__(self, max_val) -> None:
```
Comment (edgarriba) [inline kornia/losses/psnr.py]: typing

File: kornia/losses/psnr.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import torch
+import torch.nn as nn
+from torch.nn.functional import mse_loss
+
+
+class PSNR(nn.Module):
+    r"""Creates a criterion that calculates the PSNR between 2 images.
+
```
Comment (edgarriba) [inline kornia/losses/psnr.py]: can you add the formula here ? :D

File: test/test_losses.py [hunk 5]
```diff
@@ -340,3 +340,53 @@ def test_tv_on_invalid_dims(self, input):
...
+
+    def test_shape(self):
+        # Expecting an exception
+        # since we pass tensors of different shapes
+        criterion = kornia.losses.PSNR(1.0)
+        with pytest.raises(Exception) as e:
+            criterion(torch.rand(2, 3, 3, 2), torch.rand(2, 3, 3))
+
+    def test_simple(self):
+        assert pytest.approx(kornia.losses.psnr(torch.ones(1), 1.2 * torch.ones(1), 2).item(), 20.0)
```
Comment (edgarriba) [inline test/test_losses.py]: use `assert_allclose`

File: test/test_losses.py [hunk 6]
```diff
@@ -340,3 +340,53 @@ def test_tv_on_invalid_dims(self, input):
...
+
+    @pytest.mark.skip(reason="TODO: implement me")
+    def test_jit(self):
+        pass
+
+    def test_gradcheck(self):
+        signal = torch.rand(2, 3, 3, 2).double()
```
Comment (edgarriba) [inline test/test_losses.py]: `tensor_to_gradcheck_var` already does that for you

File: test/test_losses.py [hunk 7]
```diff
@@ -340,3 +340,53 @@ def test_tv_on_invalid_dims(self, input):
...
+        assert loss.shape == tuple()
+
+    def test_same_signal(self):
+        signal = torch.rand(2, 3, 3, 2)
+        approximation = signal
+
+        criterion = kornia.losses.PSNR(1.0)
+        loss = criterion(signal, approximation)
+
+        assert pytest.approx(loss.item(), float('inf'))
```
Comment (edgarriba) [inline test/test_losses.py]: use assert_allclose

**Comment by edgarriba** [comment]
@ananyamukh6  BTW, did you rebase fro mater ?

File: kornia/losses/psnr.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    r"""Function that computes PSNR
+
+    See :class:`~kornia.losses.PSNR` for details.
+    """
+    if not torch.is_tensor(signal) or not torch.is_tensor(approximation):
+        raise TypeError(f"Expected 2 torch tensors but got {type(signal)} and {type(approximation)}")
+    if signal.shape != approximation.shape:
+        raise TypeError(f"Expected tensors of equal shapes, but got {signal.shape} and {approximation.shape}")
+    mse_val = mse_loss(signal, approximation, reduction='mean')
+    return 10 * torch.log(max_val * max_val / mse_val) / torch.log(torch.tensor(10.0))
```
[the remark this answers is not in the document]
Comment (ananyamukh6) [reply kornia/losses/psnr.py]: Done

File: test/test_losses.py (hunk 5 above)
Comment (ananyamukh6) [reply test/test_losses.py]: Done

File: test/test_losses.py (hunk 7 above)
Comment (ananyamukh6) [reply test/test_losses.py]: Done

**Comment by sayantan-nervana** [comment]
> @ananyamukh6 BTW, did you rebase fro mater ?

kornia master had moved ahead, so i created a PR from kornia master to my fork's master. The current branch `ananyamukh6:ananyamukh6/psnr` is off my fork's master.

Will address the other comments in the code review during the weekend hopefully

File: kornia/losses/psnr.py (hunk 1 above)
Comment (ananyamukh6) [reply kornia/losses/psnr.py]: Done

Comment (ananyamukh6) [reply kornia/losses/psnr.py]: This is a loss between 2 tensors, `input` and `target`. Should the 3rd input be a tensor?
For example in `focal_loss` `alpha`, `gamma` and `reduction` are floats rather than tensors

File: kornia/losses/psnr.py (hunk 2 above)
Comment (ananyamukh6) [reply kornia/losses/psnr.py]: The definition of PSNR needs it to be `mean`. `sum` would be functionally incorrect, so why expose it as an user specifiable option

File: kornia/losses/psnr.py (hunk 3 above)
Comment (ananyamukh6) [reply kornia/losses/psnr.py]: Please see above comment about whether this should be a tensor or a float. Added float type for now.

File: kornia/losses/psnr.py (hunk 4 above)
Comment (ananyamukh6) [reply kornia/losses/psnr.py]: Added.

File: test/test_losses.py (hunk 6 above)
Comment (ananyamukh6) [reply test/test_losses.py]: Sorry, I do not know what it meant here. I was kind of following an existing test here: https://github.com/kornia/kornia/blob/master/test/test_losses.py#L99

**Comment by ananyamukh6** [comment]
@edgarriba , fixed the PR according to some of your comments. Added a response to some on which I need some more clarification. Thanks!

File: kornia/losses/psnr.py (hunk 1 above)
Comment (edgarriba) [reply kornia/losses/psnr.py]: it can be a float by now just to simplify things

File: test/test_losses.py (hunk 6 above)
Comment (edgarriba) [reply test/test_losses.py]: `tensor_to_gradcheck_var` cast to double precision. See: https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L36

File: kornia/losses/psnr.py (hunk 2 above)
Comment (edgarriba) [reply kornia/losses/psnr.py]: ok I see. It's fine

**Comment by edgarriba** [comment]
@ananyamukh6 please, revisit docs: https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.psnr
